File: src/main/java/cc/sukazyo/cono/morny/util/CommonEncrypt.java
package cc.sukazyo.cono.morny.util;

import javax.annotation.Nonnull;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

/**
 * 用于数据加密或编解码的工具类.
 * <p>
 * 出于 java std 中 Base64 的 {@link Base64.Encoder encode}/{@link Base64.Decoder decode} 十分好用，在此不再进行包装。
 */
public class CommonEncrypt {
	
	/**
	 * 在使用加密算法处理字符串时默认会使用的字符串编码.
	 * <p>
	 * Morny 使用 UTF-8 编码因为这是一般而言加解密工具的默认行为
	 */
	public static final Charset ENCRYPT_STANDARD_CHARSET = StandardCharsets.UTF_8;
	
	@Nonnull
	private static byte[] hashAsJavaMessageDigest(String algorithm, @Nonnull byte[] data) {
		try {
			return MessageDigest.getInstance(algorithm).digest(data);
		} catch (NoSuchAlgorithmException e) {
			throw new IllegalStateException(e);
		}
	}
	
	/**
	 * 取得数据的 md5 散列值.
	 *
	 * @param data byte 数组形式的数据体
	 * @return 二进制(byte数组)格式的数据的 md5 散列值
	 */
	@Nonnull
	public static byte[] hashMd5 (@Nonnull byte[] data) {
		return hashAsJavaMessageDigest("md5", data);
	}
	
	/**
	 * 取得一个字符串的 md5 散列值.
	 * <p>
	 * 输入的字符串将会以 {@link #ENCRYPT_STANDARD_CHARSET 默认的 UTF-8} 编码进行解析
	 *
	 * @param originString 要进行散列的字符串
	 * @return 二进制(byte数组)格式的 md5 散列值
	 */
	@Nonnull
	public static byte[] hashMd5 (String originString) {
		return hashMd5(originString.getBytes(ENCRYPT_STANDARD_CHARSET));
	}
	
	/**
	 * 取得数据的 sha1 散列值.
	 *
	 * @param data byte 数组形式的数据体
	 * @return 二进制(byte数组)格式的数据的 sha1 散列值
	 */
	@Nonnull
	public static byte[] hashSha1 (@Nonnull byte[] data) {
		return hashAsJavaMessageDigest("sha1", data);
	}
	
	/**
	 * 取得一个字符串的 sha1 散列值.
	 * <p>
	 * 输入的字符串将会以 {@link #ENCRYPT_STANDARD_CHARSET 默认的 UTF-8} 编码进行解析
	 *
	 * @param originString 要进行散列的字符串
	 * @return 二进制(byte数组)格式的 sha1 散列值
	 */
	@Nonnull
	public static byte[] hashSha1 (String originString) {
		return hashMd5(originString.getBytes(ENCRYPT_STANDARD_CHARSET));
	}
	
	/**
	 * 取得数据的 sha256 散列值.
	 *
	 * @param data byte 数组形式的数据体
	 * @return 二进制(byte数组)格式的数据的 sha256 散列值
	 */
	@Nonnull
	public static byte[] hashSha256 (@Nonnull byte[] data) {
		return hashAsJavaMessageDigest("sha256", data);
	}
	
	/**
	 * 取得一个字符串的 sha256 散列值.
	 * <p>
	 * 输入的字符串将会以 {@link #ENCRYPT_STANDARD_CHARSET 默认的 UTF-8} 编码进行解析
	 *
	 * @param originString 要进行散列的字符串
	 * @return 二进制(byte数组)格式的 sha256 散列值
	 */
	@Nonnull
	public static byte[] hashSha256 (String originString) {
		return hashMd5(originString.getBytes(ENCRYPT_STANDARD_CHARSET));
	}
	
	/**
	 * 取得数据的 sha512 散列值.
	 *
	 * @param data byte 数组形式的数据体
	 * @return 二进制(byte数组)格式的数据的 sha512 散列值
	 */
	@Nonnull
	public static byte[] hashSha512 (@Nonnull byte[] data) {
		return hashAsJavaMessageDigest("md5", data);
	}
	
	/**
	 * 取得一个字符串的 sha512 散列值.
	 * <p>
	 * 输入的字符串将会以 {@link #ENCRYPT_STANDARD_CHARSET 默认的 UTF-8} 编码进行解析
	 *
	 * @param originString 要进行散列的字符串
	 * @return 二进制(byte数组)格式的 sha512 散列值
	 */
	@Nonnull
	public static byte[] hashSha512 (String originString) {
		return hashMd5(originString.getBytes(ENCRYPT_STANDARD_CHARSET));
	}
	
	@Nonnull
	public static String base64FilenameLint (String inputName) {
		if (inputName.endsWith(".b64")) {
			return inputName.substring(0, inputName.length()-".b64".length());
		} else if (inputName.endsWith(".b64.txt")) {
			return inputName.substring(0, inputName.length()-".b64.txt".length());
		} else if (inputName.endsWith(".base64")) {
			return inputName.substring(0, inputName.length()-".base64".length());
		} else if (inputName.endsWith(".base64.txt")) {
			return inputName.substring(0, inputName.length()-".base64.txt".length());
		} else {
			return inputName;
		}
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/util/CommonEncrypt.scala
package cc.sukazyo.cono.morny.util

import java.nio.charset.{Charset, StandardCharsets}
import java.security.{MessageDigest, NoSuchAlgorithmException}

/** Provides some re-encapsulated algorithm function, and some standard values in encrypting,
  * and some normalized utils in processing something in encrypting.
  *
  * currently there's:
  *   - standard value:
  *     - [[ENCRYPT_STANDARD_CHARSET]] the standard [[Charset]] to parse between [[String]]
  *       and [[Bin]] in encrypting.
  *   - algorithm encapsulations:
  *     - [[MD5]] (MD5 Message-Digest Algorithm)
  *     - [[SHA1]] (Secure Hash Algorithm 1)
  *     - [[SHA256]] (Secure Hash Algorithm 2: 256bit)
  *     - [[SHA512]] (Secure Hash Algorithm 2: 512bit)
  *   - normalized utils
  *     - [[lint_base64FileName]] remove the .base64 file-extension for base64 text file
  *
  * @define WhenString2Bin
  * [[String]] will encoded to [[Bin]] using [[Charset]] [[ENCRYPT_STANDARD_CHARSET]]
  *
  * @todo some tests
  */
object CommonEncrypt {
	
	/** the [[Charset]] should use when converting between [[String]]
	  * and [[Bin]] in encrypting */
	val ENCRYPT_STANDARD_CHARSET: Charset = StandardCharsets.UTF_8
	
	/** the alias of [[Array]]`[`[[Byte]]`]`.
	  * means the binary data.
	  */
	//noinspection ScalaWeakerAccess
	type Bin = Array[Byte]
	
	private def hash (data: Bin)(using algorithm: String): Bin =
		try {
			MessageDigest.getInstance(algorithm) digest data
		} catch case n: NoSuchAlgorithmException =>
			throw IllegalStateException(n)
	
	/** the [[https://en.wikipedia.org/wiki/MD5 MD5]] hash value of input [[Bin]] `data`. */
	def MD5(data: Bin): Bin = hash(data)(using "md5")
	/** the [[https://en.wikipedia.org/wiki/MD5 MD5]] hash value of input [[String]] `data`.
	  *
	  * $WhenString2Bin
	  */
	def MD5 (data: String): Bin = hash(data getBytes ENCRYPT_STANDARD_CHARSET)(using "md5")
	
	/** the [[https://en.wikipedia.org/wiki/SHA-1 SHA-1]] hash value of input [[Bin]] `data`. */
	def SHA1 (data: Bin): Bin = hash(data)(using "sha1")
	/** the [[https://en.wikipedia.org/wiki/SHA-1 SHA-1]] hash value of input [[String]] `data`.
	  *
	  * $WhenString2Bin
	  */
	def SHA1 (data: String): Bin = hash(data getBytes ENCRYPT_STANDARD_CHARSET)(using "sha1")
	
	/** the [[https://en.wikipedia.org/wiki/SHA-2 SHA-2/256]] hash value of input [[Bin]] `data`. */
	def SHA256 (data: Bin): Bin = hash(data)(using "sha256")
	/** the [[https://en.wikipedia.org/wiki/SHA-2 SHA-2/256]] hash value of input [[String]] `data`.
	  *
	  * $WhenString2Bin
	  */
	def SHA256 (data: String): Bin = hash(data getBytes ENCRYPT_STANDARD_CHARSET)(using "sha256")
	
	/** the [[https://en.wikipedia.org/wiki/SHA-2 SHA-2/512]] hash value of input [[Bin]] `data`. */
	def SHA512 (data: Bin): Bin = hash(data)(using "sha512")
	/** the [[https://en.wikipedia.org/wiki/SHA-2 SHA-2/512]] hash value of input [[String]] `data`.
	  *
	  * $WhenString2Bin
	  */
	def SHA512 (data: String): Bin = hash(data getBytes ENCRYPT_STANDARD_CHARSET)(using "sha512")
	
	/** Try get the filename before it got encrypted.
	  *
	  * It assumes the base64 encrypted file should keep the original file, and plus
	  * a file-extension shows the file is base64 encrypted.
	  *
	  * Actually, the file will try find the following file-extension and drop it:
	  *   - `.b64`
	  *   - `.64.txt`
	  *   - `.base64`
	  *   - `.base64.txt`
	  * if none of those found, it will do no process anymore.
	  *
	  * @param encrypted the file fullname (means filename with file-extension) of base64 encrypted file.
	  * @return the file fullname removed the base64 file extension.
	  */
	def lint_base64FileName (encrypted: String): String = encrypted match
		case i  if i  endsWith ".b64"        => i  dropRight ".b64".length
		case ix if ix endsWith ".b64.txt"    => ix dropRight ".b64.txt".length
		case l  if l  endsWith ".base64"     => l  dropRight ".base64".length
		case lx if lx endsWith ".base64.txt" => lx dropRight ".base64.txt".length
		case u => u
	
	/** Hash a [[Long]] id to [[Bin]] using [[MD5]] algorithm.
	  *
	  * For some privacy cases, this method can provide a standard way to hash a ID to a MD5 hash value.
	  *
	  * @param id The [[Long]] number typed id.
	  * @return The hash value of the id.
	  */
	def hashId (id: Long): Bin =
		MD5(id.toString)
	
}
